src/shims/io.rs

```diff
@@ -156,15 +156,25 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
         // Remove the file handle to avoid borrowing issues
         this.remove_handle_and(fd, |mut handle, this| {
             // Don't use `?` to avoid returning before reinserting the handle
-            let bytes = this
-                .memory_mut()
-                .get_mut(buf.alloc_id).and_then(|alloc|
-                    alloc.get_bytes_mut(tcx, buf, Size::from_bytes(count))
-                    .map(|buffer| handle.file.read(buffer).map(|bytes| bytes as i64))
-                );
+
+            let bytes = this.memory_mut().get_mut(buf.alloc_id).and_then(|alloc| {
+                if count != 0 {
+                    alloc
+                        .get_bytes_mut(tcx, buf, Size::from_bytes(count))
+                        .map(|buffer| handle.file.read(buffer))
```

Comment:
Same here, shouldn't this be an and then instead of a map?

Reply:
Or just use `?`

Reply:
I think `and_then` does not work because it requires that the error variants have the same type